### PR TITLE
Allow cursed heart to be defibrillated up to three times

### PR DESCRIPTION
**Pull request by lewcc**
## What Does This PR Do
This is a different take on #18573, which simply deleted your cursed heart on death.

### Current Functionality
As it stands currently, cursed heart users are incredibly difficult to revive. The cursed heart functions by adjusting your blood level when you pump it, bringing it down below the insta-death threshold if you don't pump it as needed. 
When you get defibrillated, you get the sleeping status applied, which prevents you from manually pumping your heart, though the pump timer keeps ticking down. This means that you will almost never have enough time to get a single pump in before you instantly die again.

### *New Functionality*
When a player with a cursed heart gets defibrillated, they will be filled with enough blood to be "okay", and their heart will go into overdrive, not needing pumping for a grace period of 10 seconds. This should give them enough time to once again become conscious and start pumping their blood.
This is not the key to immortality, though; a cursed heart can only be revived this way **three times** before it will refuse to be defibrillated again, shocking whoever's trying to defibrillate them. This means you can't just have a buddy keep you alive forever like this.

## Why It's Good For The Game
I've made the argument before that this is a bit inspired by my own experiences with the cursed heart in medbay, as patients with cursed hearts tend to take a good bit of medbay's blood before a failed revive instantly deletes it all. Also, in the current state, if someone dies with a cursed heart, the only way they can be revived is by cloning or replacing their heart (which is of course limited by the 5 minute revive window). 

Instead of just deleting the heart outright, I think that this makes the experience better for the user with the cursed heart since they are no longer shocked into helplessness by the defibrillator, only to get caught in a death loop. Also, since revival requires someone else to defibrillate your dead corpse, it's not a free self-revive or anything. This also keeps a cursed heart in the game when a user dies (unlike #18573), meaning it can still be extracted and used by someone else -- with the caveat that it might not have as many revives as it originally did. 

~~Since the base cursed heart doesn't heal any damage in the first place (and it's kind of a noob trap), I think this is a fair move to make. Don't completely ruin the round of someone who picked it up off the ground. As for the wizard's heart, which heals 25 burn/brute/oxy, it only allows for a single revive. This allows wizards with apprentices a bit of flexibility; if they didn't grab a healing wand, they can still get one revive off if they're carrying around a defib. Still, this shouldn't keep the round going for too much longer, since cursed heart is enough of a handicap.~~

I've been informed that miners get the wizard heart, and so the wizard heart will be set at three revives. That should be a fair *enough* balance for an item that can heal you out of pretty much anything.

## Images of changes

https://user-images.githubusercontent.com/89928798/182257846-496e5168-c0d7-498a-b381-3ce454941aeb.mp4

## Testing
See video

## Changelog
:cl:
tweak: When being defibbed with a cursed heart, you now get a 10 second grace period before you need to start pumping manually, and you can be revived up to 3 times with it.
/:cl:

### CODE REVIEW

File: code/game/objects/items/weapons/defib.dm [hunk 1]
```diff
@@ -579,7 +588,7 @@
 					for(var/obj/item/organ/external/O in H.bodyparts)
 						total_brute	+= O.brute_dam
 						total_burn	+= O.burn_dam
-					if(total_burn <= 180 && total_brute <= 180 && !H.suiciding && !ghost && tplus < tlimit && !HAS_TRAIT(H, TRAIT_HUSK))
+					if(total_burn <= 180 && total_brute <= 180 && !H.suiciding && !ghost && tplus < tlimit && !HAS_TRAIT(H, TRAIT_HUSK) && signal_result != COMPONENT_BLOCK_DEFIB)
```
Comment (hal9000PR) [inline code/game/objects/items/weapons/defib.dm:591]: this should be a bitwise check

**Review by hal9000PR — CHANGES_REQUESTED** [review]

**Review by hal9000PR — APPROVED** [review]

File: code/__DEFINES/dcs/signals.dm [hunk 2]
```diff
@@ -390,6 +390,14 @@
 #define COMSIG_LIVING_SHOCK_PREVENTED "living_shock_prevented"
 ///sent by stuff like stunbatons and tasers: ()
 #define COMSIG_LIVING_MINOR_SHOCK "living_minor_shock"
+///Sent from defibrillators when everything seems good and the user will be shocked: (defibber, defib_item, ghost)
+#define COMSIG_LIVING_BEFORE_DEFIB "living_pre_defib"
```
Comment (SteelSlayer) [inline code/__DEFINES/dcs/signals.dm:394]: Nitpick but I think this sounds better, and falls in line with other "pre" signals.
```suggestion
#define COMSIG_LIVING_PRE_DEFIB "living_pre_defib"
```

File: code/modules/surgery/organs/heart.dm [hunk 3]
```diff
@@ -111,9 +121,68 @@
 	..()
 	if(owner)
 		to_chat(owner, "<span class='userdanger'>Your heart has been replaced with a cursed one, you have to pump this one manually otherwise you'll die!</span>")
+		RegisterSignal(owner, COMSIG_LIVING_BEFORE_DEFIB, .proc/on_defib)
+		RegisterSignal(owner, COMSIG_LIVING_DEFIBBED, .proc/on_defib_revive)
+
+/obj/item/organ/internal/heart/cursed/remove(mob/living/carbon/M, special)
+	if(owner)
+		UnregisterSignal(owner, COMSIG_LIVING_BEFORE_DEFIB)
+		UnregisterSignal(owner, COMSIG_LIVING_DEFIBBED)
```
Comment (SteelSlayer) [inline code/modules/surgery/organs/heart.dm:130]: Unregisters should be taken care of on the datum Destroy level.

**Review by AffectedArc07 — DISMISSED** [review]
Please handle steel's requested changes

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**